src/sunmatrix/sparse/sunmatrix_sparse.c

```diff
@@ -656,30 +655,38 @@ int SUNMatScaleAddI_Sparse(realtype c, SUNMatrix A)
     /* iterate through columns (rows) backwards */
     for (j=N-1; j>=0; j--) {
 
-      /* clear out temporary arrays for this column (row) */
-      for (i=0; i<M; i++) {
-        w[i] = 0;
-        x[i] = RCONST(0.0);
-      }
+      /* reset diagonal entry, in case it's not in A */
+      x[j] = ZERO;
 
       /* iterate down column (row) of A, collecting nonzeros */
-      for (p=Ap[j]; p<cend; p++) {
-        w[Ai[p]] += 1;         /* indicate that row (column) is filled */
+      for (p=Ap[j], i=0; p<cend; p++, i++) {
+        w[i] = Ai[p];        /* collect row (column) index */
         x[Ai[p]] = c*Ax[p];    /* collect/scale value */
       }
 
+      /* NNZ in this column (row) */
+      nw = cend - Ap[j];
+
       /* add identity to this column (row) */
       if (j < M) {
-        w[j] += 1;     /* indicate that row (column) is filled */
         x[j] += ONE;   /* update value */
       }
 
       /* fill entries of A with this column's (row's) data */
-      for (i=M-1; i>=0; i--) {
-        if ( w[i] > 0 ) {
-          Ai[--nz] = i;
-          Ax[nz] = x[i];
-        }
+      /* fill entries past diagonal */
+      for (i=nw-1; i>=0 && w[i]>j; i--) {
+        Ai[--nz] = w[i];
+        Ax[nz] = x[w[i]];
+      }
+      /* fill diagonal if applicable */
+      if (w[i] != j) {
```

Comment:
When there are no nonzeros in the column (row), then `nw` would be `0` yielding `i = -1` which will result in out-of-bounds indexing of `w` here.  This might be related to why the tests are failing.

Reply:
Indeed. Thank you!